### PR TITLE
LLM Scheduler

### PR DESCRIPTION
**Pull request by mahyarsamani**

### CODE REVIEW

File: src/mem/mem_scheduler.hh [hunk 1]
```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2017 Jason Lowe-Power
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:2]: Needs to be fixed

File: src/mem/mem_scheduler.hh [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+#include "params/MemScheduler.hh"
+// #include "sim/sim_object.hh"
+#include "sim/clocked_object.hh"
+#include "base/addr_range_map.hh"
+
+/**
+ * A very simple memory object. Current implementation doesn't even cache
+ * anything it just forwards requests and responses.
+ * This memobj is fully blocking (not non-blocking). Only a single request can
+ * be outstanding at a time.
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:45]: Probably should update :)

File: src/mem/mem_scheduler.hh [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+ */
+// typedef std::deque<MemPacket*> MemPacketQueue;
+class MemScheduler : public ClockedObject
+{
+  private:
+
+    /**
+     * Port on the CPU-side that receives requests.
+     * Mostly just forwards requests to the owner.
+     * Part of a vector of ports. One for each CPU port (e.g., data, inst)
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:55]: Reviewing for me, and understanding for you in a few months, is really hard when there's comments that don't match the code.

File: src/mem/mem_scheduler.hh [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+     */
+
+    struct RequestQueue{
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:58]: General comments on this struct:

1. The style needs to be fixed. We use 4 spaces to indent.
2. All of these members need to have comments to explain what they are.
3. I think this is complex enough that it should be a class, not a struct. I.e., some of the members (most of them?) should be private.
4. Does it make sense for this to be part of the CPU port???

File: src/mem/mem_scheduler.hh [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        bool _blocked;
+        /// If we tried to send a packet and it was blocked, store it here
+        PacketPtr blockedPacket;
+        const PortID _portId;
+      public:
+        /**
+         * Constructor. Just calls the superclass constructor.
+         */
+        MemSidePort(const std::string& name, PortID portId, MemScheduler *owner) :
+            RequestPort(name, owner), owner(owner), _blocked(false), blockedPacket(nullptr), _portId(portId)
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:215]: line length... How are you bypassing gem5's style checker?! ;)

File: src/mem/mem_scheduler.hh [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+         * all of the flow control is hanled in this function.
+         *
+         * @param packet to send.
+         */
+        void trySendRetry();
+
+        void sendPacket(PacketPtr pkt);
+
+        bool blocked(){
+          return _blocked;
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:229]: Like so many places... this should be 4 spaces.

File: src/mem/mem_scheduler.cc [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+    nMemPorts(params.port_mem_side_connection_count),
+    nCpuPorts(params.port_cpu_side_connection_count)
+{
+
+    panic_if(readBufferSize == 0, "readBufferSize should be non-zero");
+    panic_if(writeBufferSize == 0, "writeBufferSize "
+                                    "should be non-zero");
+    for (uint32_t i = 0; i < nCpuPorts; ++i){
+        cpuPorts.emplace_back(name() + ".cpu_side" + std::to_string(i), i, this);
+        requestQueues.push_back(RequestQueue(readBufferSize, writeBufferSize, i));
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:50]: Why push_back and not emplace? It's probably functionally the same, but push_back is creating a copy.

**Review by powerjg — COMMENTED** [review]
Looks pretty good. I have a mix of comments on the modeling aspect and code style. 

Overall, you've done a very good job! However, I want to take the chance to reiterate something that's important.

One of the values of this group is reproducible and extensible scientific contributions. One aspect of this is making the code you work on open source and/or contributing upstream. This is not possible if you do not follow the style guidelines. I suggest just getting in the habit of *never* writing any code that doesn't follow the style of the code around it. It takes a bit to learn the style, but it shouldn't take you any longer to write the actual code. By doing it when you're writing the code the first time, you save the time of the code reviewers asking you to make updates.

The same goes with writing detailed comments. However, with comments it's not just for your reviewer, but for **you**. When working on a research project, *you* will be working on this code for at a minimum 1 year. You won't be constantly modifying every line. You must write good documentation so *you* can remember why you did something 6, 12, or more, months ago.

File: src/mem/mem_scheduler.cc [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+    ADD_STAT(avgRQDelay, "Average queueing delay read requests have experienced in the MemScheduler"),
+    ADD_STAT(avgWQDelay, "Average queueing delay write requests have experienced in the MemScheduler")
+{}
+
+void
+MemScheduler::MemSchedulerStat::regStats(){
+    avgRQDelay.precision(2);
+    avgWQDelay.precision(2);
+    avgRQDelay = totalRQDelay / readReqs;
+    avgWQDelay = totalWQDelay / writeReqs;
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:195]: This can be done in the ADD_STAT macro

File: src/cpu/testers/traffic_gen/linear_gen.cc [hunk 9]
```diff
@@ -71,7 +71,7 @@ LinearGen::getNextPacket()
                               isRead ? MemCmd::ReadReq : MemCmd::WriteReq);
 
     // increment the address
-    nextAddr += blocksize;
+    nextAddr += blocksize * 16;
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/linear_gen.cc:74]: Why this change?

File: src/mem/DRAMInterface.py [hunk 10]
```diff
@@ -1414,3 +1414,184 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
     tCCD_L = "0ns"
     tRRD_L = "0ns"
     tWTR_L = "0ns"
+
+class LLM2(DRAMInterface):
+
+    beats_per_clock = 16
```
[the remark this answers is not in the document]
Comment (powerjg) [reply src/mem/DRAMInterface.py:1420]: I still think it's very important for you all to address the comments on this file.

File: src/mem/MemScheduler.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017 Jason Lowe-Power
```
[the remark this answers is not in the document]
Comment (powerjg) [reply src/mem/MemScheduler.py:2]: Still needs to be fixed.

File: src/mem/MemScheduler.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    cpu_side = VectorResponsePort("CPU side port, receives requests")
+    mem_side = VectorRequestPort("Ports on the memory "
+                                        "controllers side")
+    read_buffer_size = Param.Int(1, "Size of the internal read buffer")
+    write_buffer_size = Param.Int(1, "Size of the internal "
+                                    "write buffer")
+    resp_buffer_size = Param.Int(8, "Size of the internal "
+                                    "response buffer")
+    service_write_threshold = Param.Percent(60, "Threshold to force writes")
+    unified_queue = Param.Bool(False, "Use same queue for reads and "
```
Comment (powerjg) [inline src/mem/MemScheduler.py:45]: If this is true, what is the size of the queue? The read buffer size, write buffer size, the sum?

File: src/mem/mem_interface.cc [hunk 13]
```diff
@@ -289,11 +289,13 @@ DRAMInterface::activateBank(Rank& rank_ref, Bank& bank_ref,
     // verify that we have command bandwidth to issue the activate
     // if not, shift to next burst window
     Tick act_at;
-    if (twoCycleActivate)
+    if (twoCycleActivate){
```
Comment (powerjg) [inline src/mem/mem_interface.cc:292]: Ugh... this is why you should always use curly braces! (I'm just pointing out the poor coding practice that was already there.)

You're missing a space. (Below as well)

File: src/mem/mem_interface.cc [hunk 14]
```diff
@@ -504,15 +510,17 @@ DRAMInterface::doBurstAccess(MemPacket* mem_pkt, Tick next_burst_at,
...
     // the command; need to ensure minimum bus delay requirement is met
+    DPRINTF(DRAM, "doBurstAccess: col_allowed_at: %d, next_burst_at: %d\n", col_allowed_at, next_burst_at);
     Tick cmd_at = std::max({col_allowed_at, next_burst_at, curTick()});
 
     // verify that we have command bandwidth to issue the burst
     // if not, shift to next burst window
     if (dataClockSync && ((cmd_at - rank_ref.lastBurstTick) > clkResyncDelay))
         cmd_at = ctrl->verifyMultiCmd(cmd_at, maxCommandsPerWindow, tCK);
-    else
+    else{
```
Comment (powerjg) [inline src/mem/mem_interface.cc:520]: space
```suggestion
    else {
```

File: src/mem/mem_interface.hh [hunk 15]
```diff
@@ -924,7 +924,7 @@ class DRAMInterface : public MemInterface
     /*
      * @return time to offset next command
      */
-    Tick commandOffset() const override { return (tRP + tRCD); }
+    Tick commandOffset() const override { return (tRCD); }
```
Comment (powerjg) [inline src/mem/mem_interface.hh:927]: Is this a bug?!?!

Why this change?

File: src/mem/mem_scheduler.hh (hunk 2 above)
Comment (powerjg) [reply src/mem/mem_scheduler.hh:46]: Please update all of the comments.

File: src/mem/mem_scheduler.cc [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+    } else{
+        stats.writeReqs++;
+        entryTimes[pkt] = curTick();
+        DPRINTF(MemScheduler, "handleRequest: Pushing write request to writeQueues[%d], size = %d\n", portId, queue->writeQueue.size());
+        queue->push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: Pushed write request to writeQueues[%d], size = %d\n", portId, queue->writeQueue.size());
+    }
+    if (!nextReqEvent.scheduled()){
+        DPRINTF(MemScheduler, "handleRequest: Scheduling nextReqEvent in handleRequest\n");
+        schedule(nextReqEvent, curTick() + 100);
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:240]: What is this 100???

File: src/mem/mem_scheduler.cc [hunk 17]
```diff
@@ -0,0 +1,634 @@
...
+    DPRINTF(MemScheduler, "processNextReqEvent: Finding the least recently visited non-empty queue\n");
+    for (auto &it : requestQueues){
+        if ((it.timesChecked < minCheck) && (!it.emptyRead() || it.serviceWrite())){
+            minCheck = it.timesChecked;
+            queue = &it;
+        }
+    }
+
+    stats.totalArbitrations++;
+    DPRINTF(MemScheduler, "processNextReqEvent: Least recently visited queue found, cpuPortId = %d, (timesChecked = %d) == (minChecked = %d), readQueue.size = %d, writeQueue.size = %d, emptyRead = %d, serviceWrite = %d\n", queue->cpuPortId, queue->timesChecked, minCheck, queue->readQueue.size(), queue->writeQueue.size(), queue->emptyRead(), queue->serviceWrite());
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:259]: This line is ridiculous. 80 chars might be small, but there has to be some limit. You really should be following gem5's style guideline to a "T" even if you're working on research.

How can anyone read/understand this line?

File: src/mem/mem_scheduler.cc [hunk 18]
```diff
@@ -0,0 +1,634 @@
...
+    PacketPtr pkt;
+    if (!queue->serviceWrite()){
+        DPRINTF(MemScheduler, "processNextReqEvent: Servicing read request now\n");
+        pkt = queue->readQueue.front();
+    }
+    else{
+        DPRINTF(MemScheduler, "processNextReqEvent: Servicing write request now\n");
+        pkt = queue->writeQueue.front();
+    }
+    AddrRange addr_range = pkt->getAddrRange();
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:270]: Why range and not the address?

File: src/mem/mem_scheduler.cc [hunk 19]
```diff
@@ -0,0 +1,634 @@
...
+    DPRINTF(MemScheduler, "processNextReqEvent: pkt->addr_range: %s\n", addr_range.to_string());
+    PortID memPortId = memPortMap.contains(addr_range)->second;
+    DPRINTF(MemScheduler, "processNextReqEvent: Looked up outgoing routing table for MemSidePort PortID: %d\n", memPortId);
+    auto memPort = find_if(memPorts.begin(), memPorts.end(), [memPortId](MemSidePort &obj){return obj.portId() == memPortId;});
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:274]: It looks like memPorts should be a map instead of whatever it is. `find_if` is a strange thing to use here.

File: src/mem/mem_scheduler.cc [hunk 20]
```diff
@@ -0,0 +1,634 @@
...
+
+void 
+MemScheduler::processNextReqEventOpt(){
+    std::map<PortID, bool> visited;
+    for (auto queue : requestQueues){
+        visited[queue.cpuPortId] = false;
+    }
+    auto pick = arbitrate(visited);
+    
+    if (pick == NULL){
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:377]: use `nullptr` instead of `NULL`

**Review by powerjg — CHANGES_REQUESTED** [review]
I *advise* you to take an hour or two and go through all of my comments here and address them. I don't think it will take that long and I think it's important.

Generally, in mem_scheduler.cc I'm having a really hard time reading and reviewing the code because you have completely ignored the style guide. It's *very important* to follow the style guide of whatever code you're working on. As a researcher (and especially in this group) you are *never* writing code only for yourself. You're always writing code that others should be able to understand and contribute to. You must follow the style guide to make it feasible for others to read and understand and contribute to your code.

File: src/cpu/testers/traffic_gen/linear_gen.cc (hunk 9 above)
Comment (powerjg) [reply src/cpu/testers/traffic_gen/linear_gen.cc:74]: Ping

File: src/cpu/testers/traffic_gen/random_gen.cc [hunk 21]
```diff
@@ -61,10 +61,11 @@ RandomGen::getNextPacket()
...
 
     // address of the request
-    Addr addr = random_mt.random(startAddr, endAddr - 1);
-
+    Addr addr = random_mt.random(startAddr / blocksize, endAddr / blocksize - 1);
+    DPRINTF(TrafficGen, "startAddr: %u, endAddr: %u\n", startAddr, endAddr);
     // round down to start address of block
-    addr -= addr % blocksize;
+    // addr -= addr % blocksize;
+    addr *= blocksize;
```
Comment (powerjg) [inline src/cpu/testers/traffic_gen/random_gen.cc:68]: Probably should revert this.

File: src/mem/DRAMInterface.py [hunk 22]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
     tCCD_L = "0ns"
     tRRD_L = "0ns"
     tWTR_L = "0ns"
+
+class LLM2(DRAMInterface):
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1418]: lol, I like that you're starting with "2". Who needs a 1.0?!

File: src/mem/DRAMInterface.py [hunk 23]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
...
+
+    beats_per_clock = 64
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1420]: Can you add a comment here that says "SERDES running at 32 GHz" assuming that is what `64` beats means.

File: src/mem/DRAMInterface.py [hunk 24]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
...
+    device_bus_width = 1
+
+    # Using close-page  policy to  have deterministic
+    # memory access latency
+    page_policy = 'close'
+
+    # Total number of pins are 1024
+    # There are 1 pin/channel
+    # Atom size =  32Byte , BL= 64/1 (Bytes)
+    burst_length = 512
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1435]: This confuses me a bit.  This is the number of beats needed for 64 bytes of data? Since it's fully serialized, you need 64*8=512 beats?

Putting this in the comment would be helpful!

File: src/mem/DRAMInterface.py [hunk 25]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
     tCCD_L = "0ns"
     tRRD_L = "0ns"
     tWTR_L = "0ns"
+
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1417]: This should be in a new file (LLMInterface.py)

In fact, there's no need to put this in the `src/` directory either. All of this is configurable at runtime.

File: src/mem/DRAMInterface.py [hunk 26]
```diff
@@ -1414,3 +1414,184 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
...
+    tRCD = '15ns'
+    tCL = '7ns'
+    tRAS = '29ns'
+    tCCD_L = '16ns'
+    #tCCD_S = '2ns'
+    tBURST_MIN = '2ns'#tCCD_S = '2ns'
+    tBURST_MAX = '4ns'
+    # BL2 and BL4 supported, default to BL4
+    # DDR @ 500 MHz means 4 * 2ns / 2 = 4ns
+    tBURST = '4ns'
```
[the remark this answers is not in the document]
Comment (powerjg) [reply src/mem/DRAMInterface.py:1475]: 512 beats / 64 beats per clock = 8 clock cycles @ 500 MHz is 16ns, right??

Please update the comments above to make sense with your device. Don't quote DDR unless you have to.

File: src/mem/DRAMInterface.py [hunk 27]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
...
+
+    # value for 2Gb device from JEDEC spec
+    tRFC = '160ns'
+
+    # value for 2Gb device from JEDEC spec
+    tREFI = '3.9us'
+
+
+    tWR = '18ns'
+    # wewill mask this we consecutive requests
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1485]: I don't know what this means.

File: src/mem/DRAMInterface.py [hunk 28]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
...
+    tRTP = '0ns'
+    tWTR = '0ns'
+    tWTR_L = '0ns'
+
+    # start with 2 cycles turnaround, similar to other memory classes
+    # could be more with variations across the stack
+    tRTW = '0ns'
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1492]: This comment doesn't match the value.

File: src/mem/DRAMInterface.py [hunk 29]
```diff
@@ -1414,3 +1414,97 @@ class LPDDR5_6400_1x16_8B_BL32(LPDDR5_6400_1x16_BG_BL32):
...
+    # single rank device, set to 0
+    tCS = '0ns'
+
+    # from MemCon example, tRRD is 4ns with 2ns tCK
```
Comment (powerjg) [inline src/mem/DRAMInterface.py:1496]: What is "MemCon example"?

File: src/mem/mem_interface.cc [hunk 30]
```diff
@@ -683,6 +699,10 @@ DRAMInterface::doBurstAccess(MemPacket* mem_pkt, Tick next_burst_at,
         stats.totMemAccLat += mem_pkt->readyTime - mem_pkt->entryTime;
         stats.totQLat += cmd_at - mem_pkt->entryTime;
         stats.totBusLat += tBURST;
+
+        if (mem_pkt->readyTime - mem_pkt->entryTime > 24){
+            stats.bankConflicts++;
+        }
```
Comment (powerjg) [inline src/mem/mem_interface.cc:705]: What is this magic 24 number?

File: src/mem/mem_interface.hh (hunk 15 above)
Comment (powerjg) [reply src/mem/mem_interface.hh:930]: Ping

File: src/mem/mem_scheduler.hh [hunk 31]
```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (c) 2020 UC regents
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:2]:
```suggestion
 * Copyright (c) 2020 The Regents of the University of California.
```

File: src/mem/mem_scheduler.hh [hunk 32]
```diff
@@ -0,0 +1,398 @@
...
+#define __MEM_MEM_SCHEDULER_HH__
+
+#include <queue>
+#include <unordered_map>
+
+#include "base/statistics.hh"
+#include "mem/port.hh"
+#include "params/MemScheduler.hh"
+
+// #include "sim/sim_object.hh"
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:39]: Don't just comment things out.

Before you push any commits, run `git diff` or `git log -n1 --patch` or `git diff --cached` and read every single line that you've changed. Look for style things like whether there's commented out code.

File: src/mem/mem_scheduler.hh [hunk 33]
```diff
@@ -0,0 +1,398 @@
...
+#include "base/addr_range_map.hh"
+#include "sim/clocked_object.hh"
+
+/**
+ * A very simple memory object. This object can be connected to multiple memory
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:44]: "A very simple memory object"

Does this really describe this class? Is this not a "memory scheduler to be used with LLM" or something like that?

File: src/mem/mem_scheduler.hh (hunk 4 above)
Comment (powerjg) [reply src/mem/mem_scheduler.hh:58]: Ping...

File: src/mem/mem_scheduler.hh [hunk 34]
```diff
@@ -0,0 +1,398 @@
...
+          const PortID _portId;
+      public:
+          /**
+          * Constructor. Just calls the superclass constructor.
+          */
+          MemSidePort(const std::string& name, PortID portId,
+                      MemScheduler *owner) :
+              RequestPort(name, owner), owner(owner), _blocked(false),
+                          blockedPacket(nullptr), _portId(portId)
+          {}
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:243]: Indented too many times.

File: src/mem/mem_scheduler.hh (hunk 6 above)
Comment (powerjg) [reply src/mem/mem_scheduler.hh:256]: Ping

File: src/mem/mem_scheduler.hh [hunk 35]
```diff
@@ -0,0 +1,398 @@
...
+         * port. The default implementation ignores the change and does
+         * nothing. Override this function in a derived class if the owner
+         * needs to be aware of the address ranges, e.g. in an
+         * interconnect component like a bus.
+         */
+        void recvRangeChange() override;
+    };
+
+    /**
+     * Handle the request from the CPU side
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:287]: Can you describe at a high level what is supposed to happen on a request?

File: src/mem/mem_scheduler.hh [hunk 36]
```diff
@@ -0,0 +1,398 @@
...
+
+    void wakeUp();
+
+    /// Instantiation of the CPU-side ports
+    std::vector<CPUSidePort> cpuPorts;
+
+    /// Instantiation of the memory-side port
+    std::vector<MemSidePort> memPorts;
+
+    RequestQueue* arbitrate(std::map<PortID, bool> visited);
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:335]: Should have some comments.

File: src/mem/mem_scheduler.hh [hunk 37]
```diff
@@ -0,0 +1,398 @@
...
+    std::vector<ResponseQueue> responseQueues;
+    std::unordered_map<RequestPtr, PortID> respRoutingTable;
+    std::queue<PacketPtr> requestPool;
+    AddrRangeMap<PortID, 0> memPortMap;
+
+    std::unordered_map<PacketPtr, Tick> entryTimes;
+    std::unordered_map<PacketPtr, Tick> respEntryTimes;
+    struct MemSchedulerStat : public Stats::Group
+    {
+      MemSchedulerStat(MemScheduler *parent);
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:361]: four spaces

File: src/mem/mem_scheduler.cc [hunk 38]
```diff
@@ -0,0 +1,778 @@
...
+    stats(this)
+{
+
+    panic_if(readBufferSize == 0, "readBufferSize should be non-zero");
+    panic_if(writeBufferSize == 0, "writeBufferSize "
+                                    "should be non-zero");
+    for (uint32_t i = 0; i < nCpuPorts; ++i){
+        cpuPorts.emplace_back(name() + ".cpu_side" + std::to_string(i),
+                              i, this);
+        requestQueues.push_back(RequestQueue(readBufferSize, writeBufferSize,
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:57]: probably want `emplace` instead of `push`. Push probably does a deep copy.

File: src/mem/mem_scheduler.cc [hunk 39]
```diff
@@ -0,0 +1,778 @@
...
+                            params.service_write_threshold, unifiedQueue, i));
+        responseQueues.push_back(ResponseQueue(respBufferSize, i));
```
Comment (powerjg) [inline src/mem/mem_scheduler.cc:59]: ditto

File: src/mem/mem_scheduler.cc (hunk 8 above)
Comment (powerjg) [reply src/mem/mem_scheduler.cc:228]: ping

**Comment by powerjg** [comment]
Is there any reason for me to look at this again? There are still *many* unresolved comments from previous reviews.

**Comment by mahyarsamani** [comment]
> Is there any reason for me to look at this again? There are still _many_ unresolved comments from previous reviews.

No worries, I'm actually doing a lot of cleaning up today, you can expect to see a lot of these comments resolved tomorrow!

File: src/mem/LLMInterface.py [hunk 40]
```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2020 The Regents of the University of California.
```
Comment (powerjg) [inline src/mem/LLMInterface.py:1]: This file looks great! Thanks for all of the updates! A small thing ;)
 
```suggestion
# Copyright (c) 2021 The Regents of the University of California.
```

File: src/mem/mem_interface.cc (hunk 30 above)
Comment (powerjg) [reply src/mem/mem_interface.cc:705]: ping

File: src/mem/mem_scheduler.hh (hunk 4 above)
Comment (powerjg) [reply src/mem/mem_scheduler.hh:56]: ping...

Well, at this point, the `RequestQueue` has become too complicated to implement in a header. You should put the implementation in the cc file and make it a `class` not a `struct`

File: src/mem/mem_scheduler.hh [hunk 41]
```diff
@@ -0,0 +1,396 @@
...
+      readQueueSize(rQueueSize),
+      writeQueueSize(wQueueSize),
+      writeThreshold(writePercentage),
+      cpuPortId(portId),
+      unifiedQueue(unifiedQ),
+      sendReadRetry(false),
+      sendWriteRetry(false){}
+    };
+
+    struct ResponseQueue{
```
Comment (powerjg) [inline src/mem/mem_scheduler.hh:111]: Would it make sense for this to be a `PacketQueue`?

**Comment by powerjg** [comment]
I skipped the `MemScheduler` as it doesn't seem like my previous comments have been addressed.
